### DISCUSSION TITLE
Update DataListLoader to reduce code changes ?

### DISCUSSION THREAD

**Discussion by QuanticDisaster** [pyg-team/pytorch_geometric#3404] [Ideas]
Hello,

 I would like to know if the following behavior is intended due to limitations :

While I know `DistributedDataParallel` exists in PyTorch, `DataParallel` has the advantage of requiring minimal code changes, only requiring to apply it on the model (or specific layers).

However, In PyG, using the DataParallel of PyG library necesitates to change `DataLoader` to `DataListLoader` in code (up to this point it is still good) but any try to access attributes of a batch results in an error if code is not updated  :
`'list' object has no attribute' pos'`

This forces to update all access to attributes (labels for metrics computing), data transforms and so on

In order to reduce the code changes needed, would it be possible to improve `DataListLoader` in PyG such that all methods implemented by (acces to attributes) and on (data transformations for instance) DataLoader objects return the same result  without need to change ?

An example of the problem : 

```python
from torch import Tensor, LongTensor
from torch_geometric.loader import DataLoader, DataListLoader
from torch_geometric.data import Data


data_list = [Data(pos=Tensor(2), y=LongTensor(5), batch=LongTensor([0]), ptr=Tensor(0)), 
             Data(pos=Tensor(1), y=LongTensor(1), batch=LongTensor([2]), ptr=Tensor(1)),
             Data(pos=Tensor(3), y=LongTensor(5), batch=LongTensor([1]), ptr=Tensor(0)),
             Data(pos=Tensor(3), y=LongTensor(5), batch=LongTensor([1]), ptr=Tensor(0))]

loader = DataLoader(data_list, shuffle=True, batch_size=2)

for i in loader:
    print(i.pos)
    
parallel_loader = DataListLoader(data_list, shuffle=True, batch_size=2)

for i in parallel_loader:
    print(i.pos)
```

**Comment by rusty1s** [comment]
I don't think this is possible, as the `forward` call to a `DataParallel` module will take care of splitting the data across multiple GPUs, not the data loader. However, you can access all attributes of a batch just as in regular PyG inside the model's `forward` method, see https://github.com/pyg-team/pytorch_geometric/blob/master/examples/multi_gpu/data_parallel.py#L27.
